Implement Single Transferable Voting

BertalanD opened the pull request:
description TODO

horcsinbalint committed:
Make multiple question types

BertalanD committed:
Add sortable.js dependency

horcsinbalint committed:
.

horcsinbalint committed:
.

BertalanD committed:
Display ranked choice anonymous questions

BertalanD committed:
Hook up ranked choice to general assembly UI

BertalanD committed:
.

horcsinbalint committed:
.

horcsinbalint committed:
.

horcsinbalint committed:
.

horcsinbalint committed:
.

BertalanD committed:
Update seeders

BertalanD committed:
Display ties

BertalanD committed:
Allow abstentions during ranked choice voting

BertalanD committed:
Actually use STV (and set quota)

BertalanD committed:
Localize question creation

BertalanD committed:
Voting system should not be configurable

BertalanD committed:
Localize ranked choice voting

BertalanD committed:
Refactor result viewing

BertalanD committed:
Change Condorcet to custom fork

BertalanD committed:
Ensure that ballots with no votes don't contribute to quota

viktorcsimma committed:
Adding a table detailing the rounds of an STV count

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in 6a95774 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/707

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in 09fc13d according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/708

BertalanD committed:
Fix initials for multibyte UTF-8; full name in tooltip

BertalanD committed:
Bump condorcet to 026aa984

BertalanD committed:
.

BertalanD committed:
Allow empty ranked choice votes

An empty array is equal to `null` when checked with `==`. Use identity
comparison (`===`) instead.

BertalanD committed:
Fix number staying on in abstention -> ranking -> abstention drag

BertalanD committed:
Add option to sort items by mouse click

BertalanD committed:
Update database/factories/QuestionFactory.php

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

viktorcsimma commented on app/Http/Controllers/StudentsCouncil/QuestionController.php:
```diff
@@ -0,0 +1,89 @@
+use App\Exports\UsersSheets\AnonymousQuestionsExport;
+
+/**
+ * Controls actions related to anonymous questions.
```
Is this comment up-to-date?

viktorcsimma commented on app/Models/Question.php:
```diff
@@ -65,8 +72,15 @@
     use HasFactory;
+    use SoftDeletes;
+
+    protected $fillable = ['title', 'max_options', 'opened_at', 'closed_at', 'question_type', 'results_cache'];
```
Maybe a comment for `max_options` (that explains what it means for ranking questions) would be of use later.

BertalanD replied:
The thing is that repurposing `max_options` is a hack, and we should come up with a better way to store metadata that's only relevant to a certain type of question. Anyway, I'll leave a comment for now.

viktorcsimma commented on app/Models/Question.php:
```diff
@@ -226,11 +251,12 @@ public function storeAnswers(User $user, QuestionOption|array|string $answer, ?A
             } // else it is a string
-            elseif (!$this->has_long_answers) {
+            elseif ($this->question_type != Question::TEXT_ANSWER &&
+                    $this->question_type != Question::RANKING) {
```
Oh, so is it like a text question but with a JSON as the textual answer? Or do I understand something wrong?

BertalanD replied:
Yup, exactly. A JSON string containing an array of (numeric) answer IDs.

viktorcsimma commented on app/Rules/RankingVote.php:
```diff
@@ -0,0 +1,61 @@
+        $this->question = $question;
+    }
+
+    private function isSubset($subset, $superset)
```
Nitpick: maybe this could be static.

viktorcsimma commented on app/Rules/RankingVote.php:
```diff
@@ -0,0 +1,61 @@
+        return count(array_intersect($subset, $superset)) == count($subset);
+    }
+
+    private function isArrUni($array)
```
Same here.

viktorcsimma commented on composer.json:
```diff
@@ -11,6 +11,7 @@
         "php": "^8.2",
         "graham-campbell/markdown": "^15.0",
         "intervention/image": "^3.10",
+        "julien-boudry/condorcet": "dev-stv-tie",
```
This fails for some reason in the CI; does it have a reason it takes a longer time to fix?

BertalanD replied:
Pushed https://github.com/EotvosCollegium/mars/commit/7b21bac33031c8f5fefb14e0e0a98c7667697223 to update CI to PHP 8.3, which is the version we use in production and is required by Condorcet.

viktorcsimma commented on database/migrations/2025_02_06_214931_ranking_question.php:
```diff
@@ -0,0 +1,68 @@
+            $table->integer('max_options')->nullable()->change();
+            $table->text('results_cache')->nullable();
+            $table->timestamps();
+            $table->softDeletes();
```
Where do we use soft deletes? Or does it have some logging function?

BertalanD replied:
@horcsinbalint?

horcsinbalint replied:
Raktam törlés gombot a kérdések mellé. És soft delete-tel törlöm az átláthatóság érdekében.

horcsinbalint replied:
(I put a delete button next to the questions. And I added soft delete for greater transparency.)

viktorcsimma commented on resources/lang/hu/voting.php:
```diff
@@ -13,11 +13,12 @@
     'incorrect_passcode' => 'Érvénytelen kód',
     'is_open' => 'Nyitva?',
-    'max_options' => 'Legfeljebb hány opció választható ki:',
+    'max_options' => 'Legfeljebb hány opció választható ki/Mandátumok száma:',
```
Maybe this could be a bit more self-explanatory; like 'Legfeljebb hány opció választható ki (vagy rangsorolásos kérdésnél a kiosztandó mandátumok száma)'. Same for English.

viktorcsimma commented on resources/lang/en/voting.php:
```diff
@@ -13,11 +13,12 @@
     'incorrect_passcode' => 'Incorrect passcode',
     'is_open' => 'Is open?',
-    'max_options' => 'At most how many options should be chosen?',
+    'max_options' => 'At most how many options should be chosen?/Number of seats?',
```
What about 'Maximal number of options chosen (or for ranking questions, the number of seats to be distributed)'?

viktorcsimma commented on resources/views/student-council/general-assemblies/questions/show.blade.php:
```diff
@@ -27,15 +27,24 @@
-                        <x-input.checkbox name="option[]" value="{{$option->id}}" text="{{$option->title}}" />
-                        @endif
-                    @endforeach
+                    @if($question->question_type == \App\Models\Question::TEXT_ANSWER)
+                    @include('utils.questions.text_answer', ['question' => $question])
+                    @elseif($question->question_type == \App\Models\Question::SELECTION)
+                    @include('utils.questions.selection', ['question' => $question])
+                    @elseif($question->question_type == \App\Models\Question::RANKING)
+                    @include('utils.questions.ranking', ['question' => $question])
+                    @endif
```
Maybe it's more elegant to say something like:
```php
@include('utils.questions.' + $question->question_type, ['question' => $question])
```

viktorcsimma commented on resources/views/student-council/general-assemblies/questions/show.blade.php:
```diff
@@ -68,20 +77,19 @@
-                                @if($question->hasBeenOpened())
-                                <td><b>{{$option->votes}}</b></td>
+                            @if($question->hasBeenOpened())
+                                @if($question->question_type == \App\Models\Question::TEXT_ANSWER)
+                                    @include('utils.questions.results.text_answer', ['question' => $question])
+                                @elseif($question->question_type == \App\Models\Question::SELECTION)
+                                    @include('utils.questions.results.selection', ['question' => $question])
+                                @elseif($question->question_type == \App\Models\Question::RANKING)
+                                    @include('utils.questions.results.ranking', ['question' => $question])
```
Same here.

viktorcsimma commented on resources/lang/hu/voting.php:
```diff
@@ -48,4 +49,16 @@
+        'text_answer' => 'Szöveges válasz',
+        'ranking' => 'Preferenciális szavazás',
+    ],
+    'ranking_instructions' => 'Húzd a jelöltek neveit az általad preferált sorrendben a fenti részre (legfelülre azt, akit legjobban támogatsz). Tetszőleges számú jelöltet rangsorolhatsz; hagyd a vonal alatt azokat, akiket semmiképp sem támogatsz.',
```
We should add some instructions on abstention; like 'Tartózkodáshoz hagyd az összes jelöltet a vonal alatt.' Same for English.

viktorcsimma commented on resources/lang/en/voting.php:
```diff
@@ -48,4 +49,16 @@
+    'ranking_instructions' => 'Drag the names of the candidates in your preferred order to the section above (place the one you support most at the top). You can rank any number of candidates; leave those you don\'t support below the line.',
+    'ranking_explanation' => 'Your vote will be transferred to later candidates if the first candidate has enough votes already or has no chance of winning.',
+    'seats_allocated' => 'Number of seats to be allocated: :seats',
+    'drag_to_rank' => 'Drag options here to rank them',
+    'drag_to_exclude' => 'Drag options here to exclude from ranking',
```
Nitpick: should we drag this up to keep the alphabetical order?

viktorcsimma commented on resources/lang/hu/voting.php:
```diff
@@ -48,4 +49,16 @@
+    'ranking_instructions' => 'Húzd a jelöltek neveit az általad preferált sorrendben a fenti részre (legfelülre azt, akit legjobban támogatsz). Tetszőleges számú jelöltet rangsorolhatsz; hagyd a vonal alatt azokat, akiket semmiképp sem támogatsz.',
+    'ranking_explanation' => 'A szavazatod a későbbi jelöltekre száll, ha az első jelöltnek már elegendő szavazata van, vagy ha nincs esélye nyerni.',
+    'seats_allocated' => 'Kiosztandó mandátumok száma: :seats',
+    'drag_to_rank' => 'Húzd ide az opciókat a sorrend felállításához',
+    'drag_to_exclude' => 'Húzd ide az opciókat, amikre nem kívánsz szavazni',
```
Same here.

viktorcsimma commented on resources/views/utils/questions/results/ranking_stats.blade.php:
```diff
@@ -0,0 +1,55 @@
+            <tr>
+            @foreach($stats['rounds'][1] as $optionId => $voteNumber)
+                <th @class(['winner' => in_array($optionId, $winners)])>
+                    <span title='{{ $question->options()->where('id', $optionId)->first()->title }}'>{{$question->options()->where('id', $optionId)->first()->initials()}}</span>
```
Good idea; I didn't think about that;)

viktorcsimma reviewed (commented):
Thank you very much; it is nice!

One thing: some kind of seeder for ranking questions is sorely missing; that would make testing much easier. Even if not with randomised data, but some fixed ballots.

Most of my other comments are nitpicks; check them out if you have the time.

BertalanD committed:
Merge branch 'development' into voting

viktorcsimma committed:
Prettifying arrow buttons for ranking questions

BertalanD committed:
Fix nits

BertalanD committed:
.

BertalanD commented on app/Http/Controllers/StudentsCouncil/AnonymousQuestionController.php:
```diff
@@ -193,4 +143,17 @@ public function exportAnswerSheets(Semester $semester)
+    {
+        $this->authorize('administer', AnswerSheet::class);
+
+        if($question['parent_type'] != AnswerSheet::class) {
+            abort(400);
+        }
```
probably irrrelevant with `scopeBindings()` as that ensurers the question is the child of the specified general assembly.

viktorcsimma committed:
Adding an STV question with random ballots to the seeder

viktorcsimma committed:
Further details to stats

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in 794c626 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/711

horcsinbalint committed:
Revert regression
